File: mobile/src/screens/HomeScreen.js
import React, { useState, useEffect } from 'react';
import {
  View,
  Text,
  StyleSheet,
  RefreshControl,
  ScrollView,
  ActivityIndicator,
} from 'react-native';
import { fetchStats, fetchLatestAlarms } from '../services/apiService';
import { addAlarmListener } from '../services/socketService';

export default function HomeScreen() {
  const [stats, setStats] = useState(null);
  const [recentAlarms, setRecentAlarms] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const [isRefreshing, setIsRefreshing] = useState(false);

  useEffect(() => {
    loadData();

    // Listen for real-time alarm updates
    const removeListener = addAlarmListener((newAlarm) => {
      setRecentAlarms(prev => [newAlarm, ...prev].slice(0, 5));
      // Update stats
      setStats(prev => ({
        ...prev,
        total: (prev?.total || 0) + 1
      }));
    });

    return () => removeListener();
  }, []);

  const loadData = async () => {
    try {
      const [statsData, alarmsData] = await Promise.all([
        fetchStats(),
        fetchLatestAlarms(5)
      ]);
      setStats(statsData);
      setRecentAlarms(alarmsData);
    } catch (error) {
      console.error('Error loading data:', error);
    } finally {
      setIsLoading(false);
      setIsRefreshing(false);
    }
  };

  const onRefresh = () => {
    setIsRefreshing(true);
    loadData();
  };

  if (isLoading) {
    return (
      <View style={styles.centered}>
        <ActivityIndicator size="large" color="#007AFF" />
      </View>
    );
  }

  return (
    <ScrollView
      style={styles.container}
      refreshControl={
        <RefreshControl refreshing={isRefreshing} onRefresh={onRefresh} />
      }
    >
      <View style={styles.statsContainer}>
        <View style={styles.statCard}>
          <Text style={styles.statValue}>{stats?.total || 0}</Text>
          <Text style={styles.statLabel}>Total Alarms</Text>
        </View>
        <View style={styles.statCard}>
          <Text style={styles.statValue}>{stats?.sent_to_app || 0}</Text>
          <Text style={styles.statLabel}>Sent to App</Text>
        </View>
      </View>

      <View style={styles.sourcesContainer}>
        <Text style={styles.sectionTitle}>Alarm Sources</Text>
        {stats?.by_source && Object.entries(stats.by_source).map(([source, count]) => (
          <View key={source} style={styles.sourceRow}>
            <View style={[styles.sourceDot, { backgroundColor: getSourceColor(source) }]} />
            <Text style={styles.sourceName}>{formatSourceName(source)}</Text>
            <Text style={styles.sourceCount}>{count}</Text>
          </View>
        ))}
      </View>

      <View style={styles.recentContainer}>
        <Text style={styles.sectionTitle}>Recent Alarms</Text>
        {recentAlarms.length === 0 ? (
          <Text style={styles.emptyText}>No alarms yet</Text>
        ) : (
          recentAlarms.map((alarm) => (
            <View key={alarm.id} style={styles.alarmCard}>
              <View style={styles.alarmHeader}>
                <View style={[styles.sourceBadge, { backgroundColor: getSourceColor(alarm.source) }]}>
                  <Text style={styles.sourceBadgeText}>{formatSourceName(alarm.source)}</Text>
                </View>
                <Text style={styles.alarmTime}>
                  {formatTime(alarm.received_at)}
                </Text>
              </View>
              <Text style={styles.alarmMessage}>{alarm.message}</Text>
            </View>
          ))
        )}
      </View>
    </ScrollView>
  );
}

const getSourceColor = (source) => {
  const colors = {
    serial: '#ff6b6b',
    tap: '#4ecdc4',
    serial_ip: '#ffe66d',
  };
  return colors[source] || '#999';
};

const formatSourceName = (source) => {
  const names = {
    serial: 'Serial',
    tap: 'TAP',
    serial_ip: 'Serial/IP',
  };
  return names[source] || source;
};

const formatTime = (timestamp) => {
  const date = new Date(timestamp);
  return date.toLocaleTimeString();
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f5f5f5',
  },
  centered: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  statsContainer: {
    flexDirection: 'row',
    padding: 15,
    gap: 15,
  },
  statCard: {
    flex: 1,
    backgroundColor: '#fff',
    borderRadius: 12,
    padding: 20,
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 4,
    elevation: 3,
  },
  statValue: {
    fontSize: 32,
    fontWeight: 'bold',
    color: '#007AFF',
  },
  statLabel: {
    fontSize: 14,
    color: '#666',
    marginTop: 5,
  },
  sourcesContainer: {
    margin: 15,
    backgroundColor: '#fff',
    borderRadius: 12,
    padding: 15,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 4,
    elevation: 3,
  },
  sectionTitle: {
    fontSize: 18,
    fontWeight: 'bold',
    marginBottom: 15,
    color: '#333',
  },
  sourceRow: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 10,
    borderBottomWidth: 1,
    borderBottomColor: '#f0f0f0',
  },
  sourceDot: {
    width: 12,
    height: 12,
    borderRadius: 6,
    marginRight: 10,
  },
  sourceName: {
    flex: 1,
    fontSize: 16,
    color: '#333',
  },
  sourceCount: {
    fontSize: 16,
    fontWeight: '600',
    color: '#666',
  },
  recentContainer: {
    margin: 15,
    marginTop: 0,
  },
  alarmCard: {
    backgroundColor: '#fff',
    borderRadius: 12,
    padding: 15,
    marginBottom: 10,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 4,
    elevation: 3,
  },
  alarmHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: 10,
  },
  sourceBadge: {
    paddingHorizontal: 10,
    paddingVertical: 4,
    borderRadius: 12,
  },
  sourceBadgeText: {
    color: '#fff',
    fontSize: 12,
    fontWeight: '600',
  },
  alarmTime: {
    fontSize: 12,
    color: '#999',
  },
  alarmMessage: {
    fontSize: 16,
    color: '#333',
  },
  emptyText: {
    textAlign: 'center',
    color: '#999',
    fontSize: 16,
    marginTop: 20,
  },
});

File: mobile/lib/screens/home_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:intl/intl.dart';
import '../services/app_state.dart';
import '../models/alarm.dart';

class HomeScreen extends StatelessWidget {
  const HomeScreen({super.key});

  Color _getSourceColor(String source) {
    switch (source) {
      case 'serial':
        return const Color(0xFFFF6B6B);
      case 'tap':
        return const Color(0xFF4ECDC4);
      case 'serial_ip':
        return const Color(0xFFFFE66D);
      default:
        return Colors.grey;
    }
  }

  String _formatSourceName(String source) {
    switch (source) {
      case 'serial':
        return 'Serial';
      case 'tap':
        return 'TAP';
      case 'serial_ip':
        return 'Serial/IP';
      default:
        return source;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Home'),
        backgroundColor: Colors.blue,
        foregroundColor: Colors.white,
      ),
      body: Consumer<AppState>(
        builder: (context, appState, child) {
          if (appState.isLoading) {
            return const Center(child: CircularProgressIndicator());
          }

          final stats = appState.stats;
          final recentAlarms = appState.alarms.take(5).toList();

          return RefreshIndicator(
            onRefresh: () => appState.loadData(),
            child: ListView(
              padding: const EdgeInsets.all(16),
              children: [
                // Stats Cards
                Row(
                  children: [
                    Expanded(
                      child: _StatCard(
                        title: 'Total Alarms',
                        value: stats?.total.toString() ?? '0',
                      ),
                    ),
                    const SizedBox(width: 16),
                    Expanded(
                      child: _StatCard(
                        title: 'Sent to App',
                        value: stats?.sentToApp.toString() ?? '0',
                      ),
                    ),
                  ],
                ),
                const SizedBox(height: 20),

                // Alarm Sources
                Card(
                  child: Padding(
                    padding: const EdgeInsets.all(16),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        const Text(
                          'Alarm Sources',
                          style: TextStyle(
                            fontSize: 18,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                        const SizedBox(height: 12),
                        if (stats?.bySource.isEmpty ?? true)
                          const Text(
                            'No sources yet',
                            style: TextStyle(color: Colors.grey),
                          )
                        else
                          ...stats!.bySource.entries.map(
                            (entry) => Padding(
                              padding: const EdgeInsets.symmetric(vertical: 8),
                              child: Row(
                                children: [
                                  Container(
                                    width: 12,
                                    height: 12,
                                    decoration: BoxDecoration(
                                      color: _getSourceColor(entry.key),
                                      shape: BoxShape.circle,
                                    ),
                                  ),
                                  const SizedBox(width: 10),
                                  Expanded(
                                    child: Text(
                                      _formatSourceName(entry.key),
                                      style: const TextStyle(fontSize: 16),
                                    ),
                                  ),
                                  Text(
                                    entry.value.toString(),
                                    style: const TextStyle(
                                      fontSize: 16,
                                      fontWeight: FontWeight.w600,
                                      color: Colors.grey,
                                    ),
                                  ),
                                ],
                              ),
                            ),
                          ),
                      ],
                    ),
                  ),
                ),
                const SizedBox(height: 20),

                // Recent Alarms
                const Text(
                  'Recent Alarms',
                  style: TextStyle(
                    fontSize: 18,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                const SizedBox(height: 12),
                if (recentAlarms.isEmpty)
                  const Center(
                    child: Padding(
                      padding: EdgeInsets.all(20),
                      child: Text(
                        'No alarms yet',
                        style: TextStyle(color: Colors.grey),
                      ),
                    ),
                  )
                else
                  ...recentAlarms.map(
                    (alarm) => _AlarmCard(
                      alarm: alarm,
                      getSourceColor: _getSourceColor,
                      formatSourceName: _formatSourceName,
                    ),
                  ),
              ],
            ),
          );
        },
      ),
    );
  }
}

class _StatCard extends StatelessWidget {
  final String title;
  final String value;

  const _StatCard({
    required this.title,
    required this.value,
  });

  @override
  Widget build(BuildContext context) {
    return Card(
      elevation: 3,
      child: Padding(
        padding: const EdgeInsets.all(20),
        child: Column(
          children: [
            Text(
              value,
              style: const TextStyle(
                fontSize: 32,
                fontWeight: FontWeight.bold,
                color: Colors.blue,
              ),
            ),
            const SizedBox(height: 5),
            Text(
              title,
              style: const TextStyle(
                fontSize: 14,
                color: Colors.grey,
              ),
              textAlign: TextAlign.center,
            ),
          ],
        ),
      ),
    );
  }
}

class _AlarmCard extends StatelessWidget {
  final Alarm alarm;
  final Color Function(String) getSourceColor;
  final String Function(String) formatSourceName;

  const _AlarmCard({
    required this.alarm,
    required this.getSourceColor,
    required this.formatSourceName,
  });

  @override
  Widget build(BuildContext context) {
    return Card(
      margin: const EdgeInsets.only(bottom: 10),
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Container(
                  padding: const EdgeInsets.symmetric(
                    horizontal: 10,
                    vertical: 4,
                  ),
                  decoration: BoxDecoration(
                    color: getSourceColor(alarm.source),
                    borderRadius: BorderRadius.circular(12),
                  ),
                  child: Text(
                    formatSourceName(alarm.source),
                    style: const TextStyle(
                      color: Colors.white,
                      fontSize: 12,
                      fontWeight: FontWeight.w600,
                    ),
                  ),
                ),
                Text(
                  DateFormat.jm().format(alarm.receivedAt),
                  style: const TextStyle(
                    fontSize: 12,
                    color: Colors.grey,
                  ),
                ),
              ],
            ),
            const SizedBox(height: 10),
            Text(
              alarm.message,
              style: const TextStyle(fontSize: 16),
            ),
          ],
        ),
      ),
    );
  }
}
